### ISSUE TITLE
The correct way to 'close' a HTTP response?

### ISSUE THREAD

**Issue by redbluetoken** [warmcat/libwebsockets#1142]
Ok, I've come across an issue where either the client or the server freezes and I can't understand why. I suspect it's due to the way I am 'closing' a HTTP response.  I have a test script that sends 5000 JSON requests to the server. The server simply returns a JSON object for each request. The script is below, as well as a hacked together example that shows the behavior. 

The code in this version has  goto try_to_reuse; as the examples on the website, and if I run the example, then most of the time the example works as expected (although something will freeze now and again). If I run this example under valgrind (any tool) then after the first JSON response there is a freeze and the client will bomb out with the error after a few seconds:

Connection error: Inactivity timeout at PostClient.pl line 20.

I've tried various combinations (removing the try_to_reuse, returning -1 after the write) but nothing seems to remove this random freezing in all cases?

Is there a way to dump out all the data in the headers to see what the client/server are saying to each other? I suspect that it has something to do with keep-alive. Any other ways to help to debug this issue?

As a side, I've noticed that when running under valigrind (--tool=callgrind) the function lws_parse takes a massive amount of time (in relation to the cost of running everything else). For these 5000 JSON requests, it's called 800,000 times and takes up 25% of the total runtime. Is this normal? 

```
use strict;
use warnings;

use Mojo::UserAgent;
my $ua = Mojo::UserAgent->new;  
for my $loop (1..5000) {
  print PostJson($ua,{cmd => 'get', key=>$loop.''})."\n";
}

sub PostJson {
  my $ua = shift;
  my $json = shift;
  #print "about to post\n";
  my $tx = $ua->post('http://localhost:8061/checkaddress' => json => $json);
  #print "return from post\n";
  if (my $res = $tx->success) { return $res->body }
  else {
    my $err = $tx->error;
    die "$err->{code} response: $err->{message}" if $err->{code};
    die "Connection error: $err->{message}";
  }
}
```

```
#include <stdlib.h>
#include <string.h>
#include <fcntl.h>
#include <sys/types.h>
#include <sys/stat.h>
#ifdef WIN32
#include <io.h>
#endif
#include <stdio.h>

struct per_session_data__post_demo {
	struct lws_spa *spa;
	char result[LWS_PRE + 512];
	int result_len;
	char filename[64];
	long file_length;
        long justHeader; //a flag we'll use know if we have sent a header or not...
};


void CreateHeaders(struct lws *wsi,void *user,char* jsonText) {
	struct per_session_data__post_demo *pss = (struct per_session_data__post_demo *)user;
	unsigned char *buffer;
	unsigned char *p, *start, *end;
	int n;
		p = (unsigned char *)pss->result + LWS_PRE;
		end = p + sizeof(pss->result) - LWS_PRE - 1;
		p += sprintf((char *)p,jsonText);
                //p += lws_snprintf((char *)p, end - p, "</body></html>");
		pss->result_len = lws_ptr_diff(p, pss->result + LWS_PRE);

		n = LWS_PRE + 1024;
		buffer = malloc(n);
		p = buffer + LWS_PRE;
		start = p;
		end = p + n - LWS_PRE - 1;

		if (lws_add_http_header_status(wsi, HTTP_STATUS_OK, &p, end))
			goto bail;

		if (lws_add_http_header_by_token(wsi, WSI_TOKEN_HTTP_CONTENT_TYPE,
				(unsigned char *)"text/html", 9, &p, end))
			goto bail;
		if (lws_add_http_header_content_length(wsi, pss->result_len, &p, end))
			goto bail;
		if (lws_finalize_http_header(wsi, &p, end))
			goto bail;

		n = lws_write(wsi, start, p - start, LWS_WRITE_HTTP_HEADERS);
		if (n < 0)
			goto bail;
		free(buffer);
return;
bail:
	free(buffer);
}

static int
callback_post_demo(struct lws *wsi, enum lws_callback_reasons reason,
		   void *user, void *in, size_t len)
{
	struct per_session_data__post_demo *pss =
			(struct per_session_data__post_demo *)user;
	unsigned char *buffer;
	unsigned char *p, *start, *end;
	int n;

	switch (reason) {


	case LWS_CALLBACK_HTTP_BODY:
                pss->justHeader = 0; //flag to toggle header...
		break;
	case LWS_CALLBACK_HTTP_BODY_COMPLETION:
		printf("LWS_CALLBACK_HTTP_BODY_COMPLETION\n");
                //OK, so we now have the full JSON object passed to us, and we send it off to be processed in a different thread once it's
                //done it will create a new JSON object that we'll need to send back to the client...
		lws_callback_on_writable(wsi);
		break;
        case LWS_CALLBACK_EVENT_WAIT_CANCELLED:
          printf("callback_http LWS_CALLBACK_EVENT_WAIT_CANCELLED\n");
          //Once the other thread has handled the JSON object and created a new one it will call lws_cancel_service which will fire this event.
          //We will be lazy and get a callback on all everything...
          lws_callback_on_writable_all_protocol( lws_get_context( wsi ), lws_get_protocol( wsi ) );
        break;
	case LWS_CALLBACK_HTTP_WRITEABLE:
            printf("LWS_CALLBACK_HTTP_WRITEABLE:\n");
            //OK, but we can ONLY create the header once we know the size of the JSON object we need to return to the client.
            //BUT we have to send the header FIRST, and then ask for another callback to Send the JSON object.
                if (pss->justHeader==0) {
                  //we first send the headers...
                  CreateHeaders(wsi,user,"{ \"id\":1 }"); //hard code a responce JSON object
                  pss->justHeader=1;
                  //we need a callback so we can send the body.
                  lws_callback_on_writable(wsi);
                  return 0;
                }
                else {  
                  //we now send the body.
		  printf("LWS_CALLBACK_HTTP_WRITEABLE: sending %d\n",pss->result_len);
		  n = lws_write(wsi, (unsigned char *)pss->result + LWS_PRE,pss->result_len, LWS_WRITE_HTTP);
		  if (n < 0) return 1;
                }
                goto try_to_reuse;
              break;
	case LWS_CALLBACK_HTTP_DROP_PROTOCOL:
		break;

	default:
		break;
	}

	return 0;

try_to_reuse:
	if (lws_http_transaction_completed(wsi))
		return -1;

	return 0;
}


enum protocols
{
	PROTOCOL_HTTP = 0,
	PROTOCOL_COUNT
};

static struct lws_protocols protocols[] =
{
	/* The first protocol must always be the HTTP handler */
	{
		"protocol-post-demo",   /* name */
		callback_post_demo, /* callback */
		sizeof(struct per_session_data__post_demo), //sizeof(struct ActualClient)
		1024,
	},
	{ NULL, NULL, 0, 0 } /* terminator */
};

static struct lws_protocol_vhost_options pvo_opt = {
     NULL,
     NULL,
     "default",
     "1"
};


static struct lws_protocol_vhost_options pvo = {
     NULL,
     &pvo_opt,
     "example-protocol",
     ""
};



int WebserverServer_Start() {
  printf("WebserverServer_Start>>>: %s\n",lws_get_library_version());
  lws_set_log_level(7,NULL);


	struct lws_context_creation_info info; 
	memset( &info, 0, sizeof(info) );

	info.port = 8061;
	info.protocols = protocols; 
	info.gid = -1;
	info.uid = -1;
        info.pvo = &pvo;

	struct lws_context *context = lws_create_context( &info );

	while( 1 )
	{
		lws_service( context, /* timeout_ms = */ 1000000 );
	}

	lws_context_destroy( context );

	return 0;
}
```

**Comment by lws-team** [comment]
- I can't remember if there has ever been a time "Freeze" turned out to be the right word for whatever was happening, I don't think so.

```
lws_service( context, /* timeout_ms = */ 1000000 );
```

If you reduce this from 1000s to, eg, 1s, or 100ms, does it change the behaviour?

 - You instrumented a bytewise parser function, if you sent it lots of headers it will get called a lot.  But it doesn't take up any real time.  It's not telling you anything.

 - Also you don't need to malloc() the header buffer in CreateHeaders().  You can just write it into a buffer on the local stack.  If it couldn't all be sent immediately, lws will take care about buffering it.

**Comment by redbluetoken** [comment]
Thanks for the feedback.

Changing the time down to 1000 and 1 does change the behavior slightly (improves things), but I still get this 'freeze' when running under Valgrind. I get persistent "Premature connection close" errors on the client side (especially when it's the first connection to the server or when running more than one client). Running under Valgrind, with timeout set to 1, and turning the logging on in the client framework this is what I get:

```
dev@ubuntu:~/websocket/mv/scripts$ perl PostClient.pl 
-- Blocking request (http://localhost:8061/checkaddress)
-- Connect 83c7521dbf51c26742b8a1ad2da5cf9d (http://localhost:8061)
-- Client >>> Server (http://localhost:8061/checkaddress)
POST /checkaddress HTTP/1.1\x0d
User-Agent: Mojolicious (Perl)\x0d
Host: localhost:8061\x0d
Accept-Encoding: gzip\x0d
Content-Length: 23\x0d
Content-Type: application/json\x0d
\x0d
{"cmd":"get","key":"1"}
-- Client >>> Server (http://localhost:8061/checkaddress)

-- Client <<< Server (http://localhost:8061/checkaddress)
HTTP/1.1 200 OK\x0d
content-type: text/html\x0d
content-length: 10\x0d
\x0d

-- Client <<< Server (http://localhost:8061/checkaddress)
{ "id":1 }
-- Blocking request (http://localhost:8061/checkaddress)
-- Reusing connection 83c7521dbf51c26742b8a1ad2da5cf9d (http://localhost:8061)
-- Client >>> Server (http://localhost:8061/checkaddress)
POST /checkaddress HTTP/1.1\x0d
Content-Type: application/json\x0d
Accept-Encoding: gzip\x0d
Content-Length: 23\x0d
Host: localhost:8061\x0d
User-Agent: Mojolicious (Perl)\x0d
\x0d
{"cmd":"get","key":"2"}
-- Client >>> Server (http://localhost:8061/checkaddress)

Connection error: Premature connection close at PostClient.pl line 26.

```
You'll notice that the first request works,  the issue is with the second call, the server doesn't seem to fire LWS_CALLBACK_HTTP_BODY for this second request? The client is reusing the connection, which I assume is OK? If I re-run the script (server still running) then the script will usually finish. It will always fail when the server is restarted. I also noticed that if I do this (see <<<<<<<<<)

```
	case LWS_CALLBACK_HTTP_WRITEABLE:
            //printf("LWS_CALLBACK_HTTP_WRITEABLE:\n");
            //OK, but we can ONLY create the header once we know the size of the JSON object we need to return to the client.
            //BUT we have to send the header FIRST, and then ask for another callback to Send the JSON object.
                if (pss->justHeader==0) {
                  //we first send the headers...
                  CreateHeaders(wsi,user,"{ \"id\":1 }"); //hard code a responce JSON object
                  pss->justHeader=1;
                  //we need a callback so we can send the body.
                  lws_callback_on_writable(wsi);
                  //return 0; <<<<<<<<<
                }
                else {  
                  //we now send the body.
		  //printf("LWS_CALLBACK_HTTP_WRITEABLE: sending %d\n",pss->result_len);
		  n = lws_write(wsi, (unsigned char *)pss->result + LWS_PRE,pss->result_len, LWS_WRITE_HTTP);
		  if (n < 0) return 1;
                  //return 0;
                }
                goto try_to_reuse;
              break;
```

Then things work, but with a 5 second (or so) delay from the server (no CPU), so am I using the call to lws_http_transaction_completed correctly in all cases? 

In terms of the parse function performance, when testing my 'proper' code (not this hacked mess), I was expecting to see the call tree dominated by libc (specifically malloc/free) which is the case when the JSON objects are sent across websockets (many of these malloc/free's can be optimized out in the future). But when running the same test against HTTP, lws_parse dominates, even head of libc, which surprised me. When you say "You instrumented a bytewise parser function", does that mean there are other ways to parse the headers?

**Comment by lws-team** [comment]
Hmm... you seem to have a lot of mysteries all at once.

> but I still get this 'freeze' when running under Valgrind. I get persistent "Premature connection close" errors on the client side (especially when it's the first connection to the server or when running more than one client). 

Where's the lws logging?  Lws has very detailed logging about what it feels is happening, enable everything and see if anything interesting happened before the 'pause'.

> //return 0; <<<<<<<<<

That isn't going to fix anything or shine any light on the actual problem.

> But when running the same test against HTTP, lws_parse dominates, even head of libc, which surprised me. 

It calls the parse function once per byte of headers.  If you measure number of function calls, yes, lots of function calls there.  If you measure cpu time (without the instrumentation overhead), very little time spent there.  So what are you 'surprised' about?

**Comment by redbluetoken** [comment]
Stupid question, how do I turn on more logging for lws? I've tried various values for lws_set_log_level, ie, 2047, and have tried OR ing all the LLL_* but I get little more than:

```
[2018/01/07 10:51:42:7487] NOTICE: libuv support not compiled in
[2018/01/07 10:51:42:7488] NOTICE: Creating Vhost 'default' port 8061, 1 protocols, IPv6 off
[2018/01/07 10:51:44:8731] NOTICE: lws_http_action: 0x25d45e0: LWSS_HTTP_BODY state set
[2018/01/07 10:51:44:8731] NOTICE: HTTP_BODY_COMPLETION
[2018/01/07 10:51:50:0078] NOTICE: lws_http_action: 0x25d45e0: LWSS_HTTP_BODY state set
[2018/01/07 10:51:50:0078] NOTICE: HTTP_BODY_COMPLETION
```

Do I need to build lws in a special way to include more logging?

In terms of the parse function, valgrind seemed to be suggesting that 25% of CPU cycles was spent in this function alone (ie, hammer the server with lots of requests). Lets leave this one for the moment, I'll create more test case once I've got basic examples working 100% of the time.

**Comment by lws-team** [comment]
As it says in READMEs/README.build.md

**NOTE5**:
To build with debug info and _DEBUG for lower priority debug messages
compiled in, use
```
    $ cmake .. -DCMAKE_BUILD_TYPE=DEBUG
```

> 25% of CPU cycles was spent in this function alone 

Webserver spends 1/4 of the transaction time parsing headers doesn't sound so wild.

**Comment by redbluetoken** [comment]
Ok, I've rebuilt with debug turned on and this is what I get from lws. My interpritation of this log (and I could be wrong) is that the server does get the second request from the client but does nothing with it? See <<<>>> where the timeout happens. The client log and the code also below. Thoughts?

```
[2018/01/08 02:22:48:3302] INFO: Initial logging level 2047
[2018/01/08 02:22:48:3360] INFO: Libwebsockets version: 2.4.0 unknown-build-hash
[2018/01/08 02:22:48:3362] INFO: IPV6 not compiled in
[2018/01/08 02:22:48:3363] INFO: libev support not compiled in
[2018/01/08 02:22:48:3365] NOTICE: libuv support not compiled in
[2018/01/08 02:22:48:3368] INFO:  LWS_DEF_HEADER_LEN    : 4096
[2018/01/08 02:22:48:3370] INFO:  LWS_MAX_PROTOCOLS     : 5
[2018/01/08 02:22:48:3371] INFO:  LWS_MAX_SMP           : 1
[2018/01/08 02:22:48:3374] INFO:  sizeof (*info)        : 448
[2018/01/08 02:22:48:3375] INFO:  SYSTEM_RANDOM_FILEPATH: '/dev/urandom'
[2018/01/08 02:22:48:3377] INFO:  HTTP2 support         : not configured
[2018/01/08 02:22:48:3382] DEBUG: _realloc: size 664: context
[2018/01/08 02:22:48:3400] INFO:  default timeout (secs): 20
[2018/01/08 02:22:48:3405] DEBUG: _realloc: size 4096: pt_serv_buf
[2018/01/08 02:22:48:3412] INFO:  Threads: 1 each 1024 fds
[2018/01/08 02:22:48:3419] INFO:  mem: context:          4760 B (664 ctx + (1 thr x 4096))
[2018/01/08 02:22:48:3423] INFO:  mem: http hdr rsvd:   28384 B (1 thr x (4096 + 3000) x 4))
[2018/01/08 02:22:48:3427] DEBUG: _realloc: size 8192: fds table
[2018/01/08 02:22:48:3429] INFO:  mem: pollfd map:       8192
[2018/01/08 02:22:48:3433] DEBUG: _realloc: size 8192: lws_lookup
[2018/01/08 02:22:48:3436] INFO:  mem: platform fd map:  8192 bytes
[2018/01/08 02:22:48:3440] INFO:  Compiled with OpenSSL support
[2018/01/08 02:22:48:3442] INFO:  SSL disabled: no LWS_SERVER_OPTION_DO_SSL_GLOBAL_INIT
[2018/01/08 02:22:48:3447] DEBUG: _realloc: size 560: create vhost
[2018/01/08 02:22:48:3459] DEBUG: _realloc: size 112: vhost-specific plugin table
[2018/01/08 02:22:48:3530] DEBUG: _realloc: size 8: same vh list
[2018/01/08 02:22:48:3533] NOTICE: Creating Vhost 'default' port 8061, 1 protocols, IPv6 off
[2018/01/08 02:22:48:3581] DEBUG: _realloc: size 344: listen wsi
[2018/01/08 02:22:48:3591] DEBUG: insert_wsi_socket_into_fds: 0x5ed9660: tsi=0, sock=8, pos-in-fds=0
[2018/01/08 02:22:48:3618] INFO:  Listening on port 8061
[2018/01/08 02:22:48:3623] INFO:  mem: per-conn:          344 bytes + protocol rx buf
[2018/01/08 02:22:48:3631] INFO:  canonical_hostname = ubuntu
[2018/01/08 02:22:48:3637] DEBUG: _realloc: size 344: event pipe wsi
[2018/01/08 02:22:48:3646] DEBUG: event pipe fd 9
[2018/01/08 02:22:48:3648] DEBUG: insert_wsi_socket_into_fds: 0x7a9e380: tsi=0, sock=9, pos-in-fds=1
[2018/01/08 02:22:48:3664] DEBUG: _lws_change_pollfd: wsi 0x7a9e380: fd 9 events 1 -> 1
[2018/01/08 02:22:48:3729] INFO: lws_protocol_init
[2018/01/08 02:22:52:4824] DEBUG: accepted new conn port 59402 on fd=11
[2018/01/08 02:22:52:4849] DEBUG: _realloc: size 344: new server wsi
[2018/01/08 02:22:52:4852] DEBUG: new wsi 0x5ed9800 joining vhost default, tsi 0
[2018/01/08 02:22:52:4863] DEBUG: lws_adopt_descriptor_vhost: new wsi 0x5ed9800, sockfd 11
[2018/01/08 02:22:52:4873] DEBUG: lws_set_timeout: 0x5ed9800: 20 secs
[2018/01/08 02:22:52:4881] DEBUG: insert_wsi_socket_into_fds: 0x5ed9800: tsi=0, sock=11, pos-in-fds=2
[2018/01/08 02:22:52:4893] INFO: lws_header_table_attach: wsi 0x5ed9800: ah (nil) (tsi 0, count = 0) in
[2018/01/08 02:22:52:4916] DEBUG: _realloc: size 3000: ah struct
[2018/01/08 02:22:52:4919] DEBUG: _realloc: size 4096: ah data
[2018/01/08 02:22:52:4930] INFO: _lws_create_ah: created ah 0x7a9e520 (size 4096): pool length 1
[2018/01/08 02:22:52:4934] DEBUG: _lws_change_pollfd: wsi 0x5ed9800: fd 11 events 1 -> 1
[2018/01/08 02:22:52:4937] INFO: lws_header_table_attach: did attach wsi 0x5ed9800: ah 0x7a9e520: count 1 (on exit)
[2018/01/08 02:22:52:4945] DEBUG: lws_set_timeout: 0x5ed9800: 10 secs
[2018/01/08 02:22:52:4950] DEBUG: Attached ah immediately
[2018/01/08 02:22:52:5015] PARSER: issuing 183 bytes to parser
[2018/01/08 02:22:52:5029] PARSER: WSI_TOKEN_NAME_PART 'P' 0x50 (mode=0) wsi->lextable_pos=0
[2018/01/08 02:22:52:5042] PARSER: WSI_TOKEN_NAME_PART 'O' 0x4F (mode=0) wsi->lextable_pos=69
[2018/01/08 02:22:52:5043] PARSER: WSI_TOKEN_NAME_PART 'S' 0x53 (mode=0) wsi->lextable_pos=82
[2018/01/08 02:22:52:5094] PARSER: WSI_TOKEN_NAME_PART 'T' 0x54 (mode=0) wsi->lextable_pos=83
[2018/01/08 02:22:52:5097] PARSER: WSI_TOKEN_NAME_PART ' ' 0x20 (mode=0) wsi->lextable_pos=84
[2018/01/08 02:22:52:5105] PARSER: known hdr 1
[2018/01/08 02:22:52:5122] PARSER: WSI_TOK_(1) '/'
[2018/01/08 02:22:52:5145] PARSER: WSI_TOK_(1) 'c'
[2018/01/08 02:22:52:5148] PARSER: WSI_TOK_(1) 'h'
[2018/01/08 02:22:52:5149] PARSER: WSI_TOK_(1) 'e'
[2018/01/08 02:22:52:5150] PARSER: WSI_TOK_(1) 'c'
[2018/01/08 02:22:52:5151] PARSER: WSI_TOK_(1) 'k'
[2018/01/08 02:22:52:5152] PARSER: WSI_TOK_(1) 'a'
[2018/01/08 02:22:52:5153] PARSER: WSI_TOK_(1) 'd'
[2018/01/08 02:22:52:5154] PARSER: WSI_TOK_(1) 'd'
[2018/01/08 02:22:52:5155] PARSER: WSI_TOK_(1) 'r'
[2018/01/08 02:22:52:5156] PARSER: WSI_TOK_(1) 'e'
[2018/01/08 02:22:52:5156] PARSER: WSI_TOK_(1) 's'
[2018/01/08 02:22:52:5157] PARSER: WSI_TOK_(1) 's'
[2018/01/08 02:22:52:5158] PARSER: WSI_TOK_(1) ' '
[2018/01/08 02:22:52:5164] PARSER: WSI_TOK_(15) 'H'
[2018/01/08 02:22:52:5166] PARSER: WSI_TOK_(15) 'T'
[2018/01/08 02:22:52:5167] PARSER: WSI_TOK_(15) 'T'
[2018/01/08 02:22:52:5168] PARSER: WSI_TOK_(15) 'P'
[2018/01/08 02:22:52:5169] PARSER: WSI_TOK_(15) '/'
[2018/01/08 02:22:52:5169] PARSER: WSI_TOK_(15) '1'
[2018/01/08 02:22:52:5170] PARSER: WSI_TOK_(15) '.'
[2018/01/08 02:22:52:5171] PARSER: WSI_TOK_(15) '1'
'2018/01/08 02:22:52:5172] PARSER: WSI_TOK_(15) '
[2018/01/08 02:22:52:5174] PARSER: *
[2018/01/08 02:22:52:5176] PARSER: WSI_TOKEN_SKIPPING_SAW_CR '
'
[2018/01/08 02:22:52:5179] PARSER: WSI_TOKEN_NAME_PART 'C' 0x43 (mode=0) wsi->lextable_pos=0
[2018/01/08 02:22:52:5180] PARSER: WSI_TOKEN_NAME_PART 'o' 0x6F (mode=0) wsi->lextable_pos=117
[2018/01/08 02:22:52:5181] PARSER: WSI_TOKEN_NAME_PART 'n' 0x6E (mode=0) wsi->lextable_pos=124
[2018/01/08 02:22:52:5182] PARSER: WSI_TOKEN_NAME_PART 't' 0x74 (mode=0) wsi->lextable_pos=131
[2018/01/08 02:22:52:5183] PARSER: WSI_TOKEN_NAME_PART 'e' 0x65 (mode=0) wsi->lextable_pos=524
[2018/01/08 02:22:52:5184] PARSER: WSI_TOKEN_NAME_PART 'n' 0x6E (mode=0) wsi->lextable_pos=525
[2018/01/08 02:22:52:5185] PARSER: WSI_TOKEN_NAME_PART 't' 0x74 (mode=0) wsi->lextable_pos=526
[2018/01/08 02:22:52:5186] PARSER: WSI_TOKEN_NAME_PART '-' 0x2D (mode=0) wsi->lextable_pos=527
[2018/01/08 02:22:52:5187] PARSER: WSI_TOKEN_NAME_PART 'T' 0x54 (mode=0) wsi->lextable_pos=528
[2018/01/08 02:22:52:5188] PARSER: WSI_TOKEN_NAME_PART 'y' 0x79 (mode=0) wsi->lextable_pos=561
[2018/01/08 02:22:52:5189] PARSER: WSI_TOKEN_NAME_PART 'p' 0x70 (mode=0) wsi->lextable_pos=562
[2018/01/08 02:22:52:5190] PARSER: WSI_TOKEN_NAME_PART 'e' 0x65 (mode=0) wsi->lextable_pos=563
[2018/01/08 02:22:52:5190] PARSER: WSI_TOKEN_NAME_PART ':' 0x3A (mode=0) wsi->lextable_pos=564
[2018/01/08 02:22:52:5191] PARSER: known hdr 28
[2018/01/08 02:22:52:5192] PARSER: WSI_TOK_(28) ' '
[2018/01/08 02:22:52:5194] PARSER: WSI_TOK_(28) 'a'
[2018/01/08 02:22:52:5194] PARSER: WSI_TOK_(28) 'p'
[2018/01/08 02:22:52:5195] PARSER: WSI_TOK_(28) 'p'
[2018/01/08 02:22:52:5196] PARSER: WSI_TOK_(28) 'l'
[2018/01/08 02:22:52:5197] PARSER: WSI_TOK_(28) 'i'
[2018/01/08 02:22:52:5198] PARSER: WSI_TOK_(28) 'c'
[2018/01/08 02:22:52:5199] PARSER: WSI_TOK_(28) 'a'
[2018/01/08 02:22:52:5200] PARSER: WSI_TOK_(28) 't'
[2018/01/08 02:22:52:5200] PARSER: WSI_TOK_(28) 'i'
[2018/01/08 02:22:52:5202] PARSER: WSI_TOK_(28) 'o'
[2018/01/08 02:22:52:5202] PARSER: WSI_TOK_(28) 'n'
[2018/01/08 02:22:52:5203] PARSER: WSI_TOK_(28) '/'
[2018/01/08 02:22:52:5204] PARSER: WSI_TOK_(28) 'j'
[2018/01/08 02:22:52:5205] PARSER: WSI_TOK_(28) 's'
[2018/01/08 02:22:52:5206] PARSER: WSI_TOK_(28) 'o'
[2018/01/08 02:22:52:5207] PARSER: WSI_TOK_(28) 'n'
'2018/01/08 02:22:52:5208] PARSER: WSI_TOK_(28) '
[2018/01/08 02:22:52:5208] PARSER: *
[2018/01/08 02:22:52:5209] PARSER: WSI_TOKEN_SKIPPING_SAW_CR '
'
[2018/01/08 02:22:52:5210] PARSER: WSI_TOKEN_NAME_PART 'A' 0x41 (mode=0) wsi->lextable_pos=0
[2018/01/08 02:22:52:5212] PARSER: WSI_TOKEN_NAME_PART 'c' 0x63 (mode=0) wsi->lextable_pos=329
[2018/01/08 02:22:52:5213] PARSER: WSI_TOKEN_NAME_PART 'c' 0x63 (mode=0) wsi->lextable_pos=342
[2018/01/08 02:22:52:5213] PARSER: WSI_TOKEN_NAME_PART 'e' 0x65 (mode=0) wsi->lextable_pos=343
[2018/01/08 02:22:52:5214] PARSER: WSI_TOKEN_NAME_PART 'p' 0x70 (mode=0) wsi->lextable_pos=344
[2018/01/08 02:22:52:5215] PARSER: WSI_TOKEN_NAME_PART 't' 0x74 (mode=0) wsi->lextable_pos=351
[2018/01/08 02:22:52:5216] PARSER: WSI_TOKEN_NAME_PART '-' 0x2D (mode=0) wsi->lextable_pos=352
[2018/01/08 02:22:52:5217] PARSER: WSI_TOKEN_NAME_PART 'E' 0x45 (mode=0) wsi->lextable_pos=444
[2018/01/08 02:22:52:5218] PARSER: WSI_TOKEN_NAME_PART 'n' 0x6E (mode=0) wsi->lextable_pos=457
[2018/01/08 02:22:52:5219] PARSER: WSI_TOKEN_NAME_PART 'c' 0x63 (mode=0) wsi->lextable_pos=458
[2018/01/08 02:22:52:5220] PARSER: WSI_TOKEN_NAME_PART 'o' 0x6F (mode=0) wsi->lextable_pos=459
[2018/01/08 02:22:52:5221] PARSER: WSI_TOKEN_NAME_PART 'd' 0x64 (mode=0) wsi->lextable_pos=460
[2018/01/08 02:22:52:5222] PARSER: WSI_TOKEN_NAME_PART 'i' 0x69 (mode=0) wsi->lextable_pos=461
[2018/01/08 02:22:52:5223] PARSER: WSI_TOKEN_NAME_PART 'n' 0x6E (mode=0) wsi->lextable_pos=462
[2018/01/08 02:22:52:5223] PARSER: WSI_TOKEN_NAME_PART 'g' 0x67 (mode=0) wsi->lextable_pos=463
[2018/01/08 02:22:52:5224] PARSER: WSI_TOKEN_NAME_PART ':' 0x3A (mode=0) wsi->lextable_pos=464
[2018/01/08 02:22:52:5225] PARSER: known hdr 21
[2018/01/08 02:22:52:5226] PARSER: WSI_TOK_(21) ' '
[2018/01/08 02:22:52:5227] PARSER: WSI_TOK_(21) 'g'
[2018/01/08 02:22:52:5228] PARSER: WSI_TOK_(21) 'z'
[2018/01/08 02:22:52:5229] PARSER: WSI_TOK_(21) 'i'
[2018/01/08 02:22:52:5230] PARSER: WSI_TOK_(21) 'p'
'2018/01/08 02:22:52:5230] PARSER: WSI_TOK_(21) '
[2018/01/08 02:22:52:5232] PARSER: *
[2018/01/08 02:22:52:5232] PARSER: WSI_TOKEN_SKIPPING_SAW_CR '
'
[2018/01/08 02:22:52:5233] PARSER: WSI_TOKEN_NAME_PART 'C' 0x43 (mode=0) wsi->lextable_pos=0
[2018/01/08 02:22:52:5234] PARSER: WSI_TOKEN_NAME_PART 'o' 0x6F (mode=0) wsi->lextable_pos=117
[2018/01/08 02:22:52:5235] PARSER: WSI_TOKEN_NAME_PART 'n' 0x6E (mode=0) wsi->lextable_pos=124
[2018/01/08 02:22:52:5236] PARSER: WSI_TOKEN_NAME_PART 't' 0x74 (mode=0) wsi->lextable_pos=131
[2018/01/08 02:22:52:5237] PARSER: WSI_TOKEN_NAME_PART 'e' 0x65 (mode=0) wsi->lextable_pos=524
[2018/01/08 02:22:52:5238] PARSER: WSI_TOKEN_NAME_PART 'n' 0x6E (mode=0) wsi->lextable_pos=525
[2018/01/08 02:22:52:5239] PARSER: WSI_TOKEN_NAME_PART 't' 0x74 (mode=0) wsi->lextable_pos=526
[2018/01/08 02:22:52:5240] PARSER: WSI_TOKEN_NAME_PART '-' 0x2D (mode=0) wsi->lextable_pos=527
[2018/01/08 02:22:52:5241] PARSER: WSI_TOKEN_NAME_PART 'L' 0x4C (mode=0) wsi->lextable_pos=528
[2018/01/08 02:22:52:5242] PARSER: WSI_TOKEN_NAME_PART 'e' 0x65 (mode=0) wsi->lextable_pos=544
[2018/01/08 02:22:52:5243] PARSER: WSI_TOKEN_NAME_PART 'n' 0x6E (mode=0) wsi->lextable_pos=554
[2018/01/08 02:22:52:5244] PARSER: WSI_TOKEN_NAME_PART 'g' 0x67 (mode=0) wsi->lextable_pos=555
[2018/01/08 02:22:52:5245] PARSER: WSI_TOKEN_NAME_PART 't' 0x74 (mode=0) wsi->lextable_pos=556
[2018/01/08 02:22:52:5245] PARSER: WSI_TOKEN_NAME_PART 'h' 0x68 (mode=0) wsi->lextable_pos=557
[2018/01/08 02:22:52:5246] PARSER: WSI_TOKEN_NAME_PART ':' 0x3A (mode=0) wsi->lextable_pos=558
[2018/01/08 02:22:52:5247] PARSER: known hdr 27
[2018/01/08 02:22:52:5248] PARSER: WSI_TOK_(27) ' '
[2018/01/08 02:22:52:5249] PARSER: WSI_TOK_(27) '2'
[2018/01/08 02:22:52:5250] PARSER: WSI_TOK_(27) '3'
'2018/01/08 02:22:52:5250] PARSER: WSI_TOK_(27) '
[2018/01/08 02:22:52:5252] PARSER: *
[2018/01/08 02:22:52:5253] PARSER: WSI_TOKEN_SKIPPING_SAW_CR '
'
[2018/01/08 02:22:52:5254] PARSER: WSI_TOKEN_NAME_PART 'H' 0x48 (mode=0) wsi->lextable_pos=0
[2018/01/08 02:22:52:5254] PARSER: WSI_TOKEN_NAME_PART 'o' 0x6F (mode=0) wsi->lextable_pos=102
[2018/01/08 02:22:52:5255] PARSER: WSI_TOKEN_NAME_PART 's' 0x73 (mode=0) wsi->lextable_pos=112
[2018/01/08 02:22:52:5257] PARSER: WSI_TOKEN_NAME_PART 't' 0x74 (mode=0) wsi->lextable_pos=113
[2018/01/08 02:22:52:5257] PARSER: WSI_TOKEN_NAME_PART ':' 0x3A (mode=0) wsi->lextable_pos=114
[2018/01/08 02:22:52:5258] PARSER: known hdr 3
[2018/01/08 02:22:52:5259] PARSER: WSI_TOK_(3) ' '
[2018/01/08 02:22:52:5260] PARSER: WSI_TOK_(3) 'l'
[2018/01/08 02:22:52:5269] PARSER: WSI_TOK_(3) 'o'
[2018/01/08 02:22:52:5271] PARSER: WSI_TOK_(3) 'c'
[2018/01/08 02:22:52:5272] PARSER: WSI_TOK_(3) 'a'
[2018/01/08 02:22:52:5273] PARSER: WSI_TOK_(3) 'l'
[2018/01/08 02:22:52:5274] PARSER: WSI_TOK_(3) 'h'
[2018/01/08 02:22:52:5274] PARSER: WSI_TOK_(3) 'o'
[2018/01/08 02:22:52:5275] PARSER: WSI_TOK_(3) 's'
[2018/01/08 02:22:52:5276] PARSER: WSI_TOK_(3) 't'
[2018/01/08 02:22:52:5277] PARSER: WSI_TOK_(3) ':'
[2018/01/08 02:22:52:5278] PARSER: WSI_TOK_(3) '8'
[2018/01/08 02:22:52:5279] PARSER: WSI_TOK_(3) '0'
[2018/01/08 02:22:52:5279] PARSER: WSI_TOK_(3) '6'
[2018/01/08 02:22:52:5280] PARSER: WSI_TOK_(3) '1'
'2018/01/08 02:22:52:5282] PARSER: WSI_TOK_(3) '
[2018/01/08 02:22:52:5283] PARSER: *
[2018/01/08 02:22:52:5283] PARSER: WSI_TOKEN_SKIPPING_SAW_CR '
'
[2018/01/08 02:22:52:5285] PARSER: WSI_TOKEN_NAME_PART 'U' 0x55 (mode=0) wsi->lextable_pos=0
[2018/01/08 02:22:52:5286] PARSER: WSI_TOKEN_NAME_PART 's' 0x73 (mode=0) wsi->lextable_pos=153
[2018/01/08 02:22:52:5288] PARSER: WSI_TOKEN_NAME_PART 'e' 0x65 (mode=0) wsi->lextable_pos=1028
[2018/01/08 02:22:52:5289] PARSER: WSI_TOKEN_NAME_PART 'r' 0x72 (mode=0) wsi->lextable_pos=1029
[2018/01/08 02:22:52:5290] PARSER: WSI_TOKEN_NAME_PART '-' 0x2D (mode=0) wsi->lextable_pos=1030
[2018/01/08 02:22:52:5292] PARSER: WSI_TOKEN_NAME_PART 'A' 0x41 (mode=0) wsi->lextable_pos=1031
[2018/01/08 02:22:52:5293] PARSER: WSI_TOKEN_NAME_PART 'g' 0x67 (mode=0) wsi->lextable_pos=1032
[2018/01/08 02:22:52:5294] PARSER: WSI_TOKEN_NAME_PART 'e' 0x65 (mode=0) wsi->lextable_pos=1033
[2018/01/08 02:22:52:5295] PARSER: WSI_TOKEN_NAME_PART 'n' 0x6E (mode=0) wsi->lextable_pos=1034
[2018/01/08 02:22:52:5296] PARSER: WSI_TOKEN_NAME_PART 't' 0x74 (mode=0) wsi->lextable_pos=1035
[2018/01/08 02:22:52:5297] PARSER: WSI_TOKEN_NAME_PART ':' 0x3A (mode=0) wsi->lextable_pos=1036
[2018/01/08 02:22:52:5298] PARSER: known hdr 69
[2018/01/08 02:22:52:5299] PARSER: WSI_TOK_(69) ' '
[2018/01/08 02:22:52:5300] PARSER: WSI_TOK_(69) 'M'
[2018/01/08 02:22:52:5301] PARSER: WSI_TOK_(69) 'o'
[2018/01/08 02:22:52:5302] PARSER: WSI_TOK_(69) 'j'
[2018/01/08 02:22:52:5303] PARSER: WSI_TOK_(69) 'o'
[2018/01/08 02:22:52:5304] PARSER: WSI_TOK_(69) 'l'
[2018/01/08 02:22:52:5305] PARSER: WSI_TOK_(69) 'i'
[2018/01/08 02:22:52:5305] PARSER: WSI_TOK_(69) 'c'
[2018/01/08 02:22:52:5307] PARSER: WSI_TOK_(69) 'i'
[2018/01/08 02:22:52:5307] PARSER: WSI_TOK_(69) 'o'
[2018/01/08 02:22:52:5308] PARSER: WSI_TOK_(69) 'u'
[2018/01/08 02:22:52:5309] PARSER: WSI_TOK_(69) 's'
[2018/01/08 02:22:52:5310] PARSER: WSI_TOK_(69) ' '
[2018/01/08 02:22:52:5312] PARSER: WSI_TOK_(69) '('
[2018/01/08 02:22:52:5312] PARSER: WSI_TOK_(69) 'P'
[2018/01/08 02:22:52:5313] PARSER: WSI_TOK_(69) 'e'
[2018/01/08 02:22:52:5314] PARSER: WSI_TOK_(69) 'r'
[2018/01/08 02:22:52:5315] PARSER: WSI_TOK_(69) 'l'
[2018/01/08 02:22:52:5316] PARSER: WSI_TOK_(69) ')'
'2018/01/08 02:22:52:5317] PARSER: WSI_TOK_(69) '
[2018/01/08 02:22:52:5318] PARSER: *
[2018/01/08 02:22:52:5319] PARSER: WSI_TOKEN_SKIPPING_SAW_CR '
'
' 0x0D (mode=0) wsi->lextable_pos=0WSI_TOKEN_NAME_PART '
[2018/01/08 02:22:52:5321] PARSER: WSI_TOKEN_NAME_PART '
' 0x0A (mode=0) wsi->lextable_pos=238
[2018/01/08 02:22:52:5322] PARSER: known hdr 8
[2018/01/08 02:22:52:5329] PARSER: lws_handshake_server: lws_parse sees parsing complete
[2018/01/08 02:22:52:5331] DEBUG: lws_handshake_server: wsi->more_rx_waiting=1
[2018/01/08 02:22:52:5350] INFO: vhost match to default based on port 8061
[2018/01/08 02:22:52:5357] DEBUG: lws_set_timeout: 0x5ed9800: 0 secs
[2018/01/08 02:22:52:5364] INFO: No upgrade
[2018/01/08 02:22:52:5369] DEBUG: lws_union_transition: 0x5ed9800: mode 1
[2018/01/08 02:22:52:5372] DEBUG: lws_handshake_server: wsi 0x5ed9800: ah 0x7a9e520
[2018/01/08 02:22:52:5387] INFO: Method: 'POST' (1), request for '/checkaddress'
[2018/01/08 02:22:52:5391] DEBUG: _realloc: size 624: user space
[2018/01/08 02:22:52:5488] DEBUG: lws_set_timeout: 0x5ed9800: 20 secs
[2018/01/08 02:22:52:5494] INFO: no hit
[2018/01/08 02:22:52:5501] INFO: lws_same_vh_protocol_remove: removing same prot wsi 0x5ed9800
[2018/01/08 02:22:52:5505] DEBUG: _realloc: size 624: user space
[2018/01/08 02:22:52:5525] DEBUG: wsi->http.rx_content_length 23 0 0
[2018/01/08 02:22:52:5528] NOTICE: lws_http_action: 0x5ed9800: LWSS_HTTP_BODY state set
[2018/01/08 02:22:52:5534] DEBUG: lws_read: thinks we have used 23
LWS_CALLBACK_HTTP_BODY
[2018/01/08 02:22:52:5545] DEBUG: lws_set_timeout: 0x5ed9800: 0 secs
[2018/01/08 02:22:52:5546] NOTICE: HTTP_BODY_COMPLETION
LWS_CALLBACK_HTTP_BODY_COMPLETION
[2018/01/08 02:22:52:5559] DEBUG: _lws_change_pollfd: wsi 0x5ed9800: fd 11 events 1 -> 5
[2018/01/08 02:22:52:5567] INFO: lws_read: 0x5ed9800: read_ok, used 183 (len 0, state 67)
[2018/01/08 02:22:52:5572] DEBUG: lws_server_socket_service: wsi 0x5ed9800: ah read rxpos 183, rxlen 183
[2018/01/08 02:22:52:5586] DEBUG: _lws_change_pollfd: wsi 0x5ed9800: fd 11 events 5 -> 1
LWS_CALLBACK_HTTP_WRITEABLE pss->justHeader = 0
[2018/01/08 02:22:52:5643] INFO: lws_add_http_header_content_length: wsi 0x5ed9800: tx_content_length/remain 10
[2018/01/08 02:22:52:5698] DEBUG: _lws_change_pollfd: wsi 0x5ed9800: fd 11 events 1 -> 5
[2018/01/08 02:22:52:5708] DEBUG: _lws_change_pollfd: wsi 0x5ed9800: fd 11 events 5 -> 1
LWS_CALLBACK_HTTP_WRITEABLE pss->justHeader = 1
[2018/01/08 02:22:52:5724] INFO: lws_http_transaction_completed: wsi 0x5ed9800
[2018/01/08 02:22:52:5727] DEBUG: lws_http_transaction_completed: wsi 0x5ed9800
[2018/01/08 02:22:52:5731] INFO: lws_same_vh_protocol_remove: removing same prot wsi 0x5ed9800
[2018/01/08 02:22:52:5733] INFO: have prev 0x7a9e330, setting him to our next (nil)
[2018/01/08 02:22:52:5735] DEBUG: _realloc: size 624: user space
[2018/01/08 02:22:52:5739] DEBUG: lws_set_timeout: 0x5ed9800: 5 secs
[2018/01/08 02:22:52:5742] DEBUG: lws_http_transaction_completed: wsi->more_rx_waiting=1
[2018/01/08 02:22:52:5745] DEBUG: lws_set_timeout: 0x5ed9800: 10 secs
[2018/01/08 02:22:52:5747] DEBUG: lws_set_timeout: 0x5ed9800: 5 secs
[2018/01/08 02:22:52:5749] INFO: lws_http_transaction_completed: 0x5ed9800: keep-alive await new transaction
[2018/01/08 02:22:52:5815] DEBUG: _lws_change_pollfd: wsi 0x5ed9800: fd 11 events 1 -> 5
[2018/01/08 02:22:52:5820] DEBUG: lws_server_socket_service: wsi 0x5ed9800 read 183
[2018/01/08 02:22:52:5825] DEBUG: lws_read: LWSS_HTTP_DEFERRING_ACTION
[2018/01/08 02:22:52:5842] INFO: lws_read: 0x5ed9800: read_ok, used 0 (len 183, state 211)
[2018/01/08 02:22:52:5847] DEBUG: _lws_change_pollfd: wsi 0x5ed9800: fd 11 events 5 -> 1
[2018/01/08 02:22:52:5848] DEBUG: lws_server_socket_service: LWSS_HTTP_DEFERRING_ACTION now writable
[2018/01/08 02:22:52:5851] DEBUG: _lws_change_pollfd: wsi 0x5ed9800: fd 11 events 1 -> 1
<<<>>>
[2018/01/08 02:22:58:0053] INFO: wsi 0x5ed9800: TIMEDOUT WAITING on 25 (did hdr 0, ah 0x7a9e520, wl 0, pfd events 1) 1515406978 vs 5
[2018/01/08 02:22:58:0058] DEBUG: lws_close_free_wsi: 0x5ed9800
[2018/01/08 02:22:58:0074] DEBUG: lws_close_free_wsi: 0x5ed9800: DROP_PROTOCOL protocol-post-demo
[2018/01/08 02:22:58:0081] DEBUG: lws_close_free_wsi: real just_kill_connection: 0x5ed9800 (sockfd 11)
[2018/01/08 02:22:58:0096] INFO: lws_same_vh_protocol_remove: removing same prot wsi 0x5ed9800
[2018/01/08 02:22:58:0097] INFO: have prev 0x7a9e330, setting him to our next (nil)
[2018/01/08 02:22:58:0100] DEBUG: remove_wsi_socket_from_fds: wsi=0x5ed9800, sock=11, fds pos=2, end guy pos=3, endfd=0
[2018/01/08 02:22:58:0199] INFO: ah det due to close
[2018/01/08 02:22:58:0208] INFO: lws_header_table_detach: wsi 0x5ed9800: ah 0x7a9e520 (tsi=0, count = 1)
[2018/01/08 02:22:58:0220] DEBUG: lws_header_table_detach: wsi 0x5ed9800: ah held 6s, ah.rxpos -1, ah.rxlen -1, mode/state 1 4,wsi->more_rx_waiting 1
[2018/01/08 02:22:58:0226] INFO: lws_header_table_detach: nobody usable waiting
[2018/01/08 02:22:58:0230] INFO: _lws_destroy_ah: freed ah 0x7a9e520 : pool length 0
[2018/01/08 02:22:58:0237] INFO: lws_header_table_detach: wsi 0x5ed9800: ah 0x7a9e520 (tsi=0, count = 0)
[2018/01/08 02:22:58:0244] DEBUG: lws_free_wsi: 0x5ed9800, remaining wsi 2
```

Client log:
```

-- Blocking request (http://localhost:8061/checkaddress)
-- Connect c346a0bebe35a6bc81337320b4227985 (http://localhost:8061)
-- Client >>> Server (http://localhost:8061/checkaddress)
POST /checkaddress HTTP/1.1\x0d
Content-Type: application/json\x0d
Accept-Encoding: gzip\x0d
Content-Length: 23\x0d
Host: localhost:8061\x0d
User-Agent: Mojolicious (Perl)\x0d
\x0d
{"cmd":"get","key":"1"}
-- Client >>> Server (http://localhost:8061/checkaddress)

-- Client <<< Server (http://localhost:8061/checkaddress)
HTTP/1.1 200 OK\x0d
content-type: text/html\x0d
content-length: 10\x0d
\x0d

-- Client <<< Server (http://localhost:8061/checkaddress)
{ "id":1 }
-- Blocking request (http://localhost:8061/checkaddress)
-- Reusing connection c346a0bebe35a6bc81337320b4227985 (http://localhost:8061)
-- Client >>> Server (http://localhost:8061/checkaddress)
POST /checkaddress HTTP/1.1\x0d
User-Agent: Mojolicious (Perl)\x0d
Host: localhost:8061\x0d
Content-Length: 23\x0d
Accept-Encoding: gzip\x0d
Content-Type: application/json\x0d
\x0d
{"cmd":"get","key":"2"}
-- Client >>> Server (http://localhost:8061/checkaddress)

Connection error: Premature connection close at PostClient.pl line 27.
```
code:

```
static int
callback_post_demo(struct lws *wsi, enum lws_callback_reasons reason,
		   void *user, void *in, size_t len)
{
	struct per_session_data__post_demo *pss =
			(struct per_session_data__post_demo *)user;
	unsigned char *buffer;
	unsigned char *p, *start, *end;
	int n;

	switch (reason) {
	case LWS_CALLBACK_HTTP_BODY:
                printf("LWS_CALLBACK_HTTP_BODY\n");
                pss->justHeader = 0; //flag to toggle header...
		break;
	case LWS_CALLBACK_HTTP_BODY_COMPLETION:
		printf("LWS_CALLBACK_HTTP_BODY_COMPLETION\n");
                //OK, so we now have the full JSON object passed to us, and we send it off to be processed in a different thread once it's
                //done it will create a new JSON object that we'll need to send back to the client...
		lws_callback_on_writable(wsi);
		break;
        case LWS_CALLBACK_EVENT_WAIT_CANCELLED:
          printf("callback_http LWS_CALLBACK_EVENT_WAIT_CANCELLED\n");
          //Once the other thread has handled the JSON object and created a new one it will call lws_cancel_service which will fire this event.
          //We will be lazy and get a callback on all everything...
          lws_callback_on_writable_all_protocol( lws_get_context( wsi ), lws_get_protocol( wsi ) );
        break;
	case LWS_CALLBACK_HTTP_WRITEABLE:
                printf("LWS_CALLBACK_HTTP_WRITEABLE pss->justHeader = %li\n",pss->justHeader);
                if (pss->justHeader==0) {
                  //we first send the headers...
                  CreateHeaders(wsi,user,"{ \"id\":1 }"); //hard code a responce JSON object
                  pss->justHeader=1;
                  //we need a callback so we can send the body.
                  lws_callback_on_writable(wsi);
                  return 0;
                }
                if (pss->justHeader==1) {
                  //we now send the body.
		  n = lws_write(wsi, (unsigned char *)pss->result + LWS_PRE,pss->result_len, LWS_WRITE_HTTP);
		  if (n < 0) return 1;
                  pss->justHeader=2; //just in case we get other LWS_CALLBACK_HTTP_WRITEABLE
                }
 	        if (lws_http_transaction_completed(wsi)) {
                  printf("lws_http_transaction_completed true\n");
		  return -1;
                }
	        return 0;
              break;
	case LWS_CALLBACK_HTTP_DROP_PROTOCOL:
        break;

	return 0;
}
```

**Comment by redbluetoken** [comment]
I don't know if this is related, a separate issue, or a lack of my understanding:

When the client is doing:

loop {
  open connection;
    Post Request; wait for response; 
  close connection;
}

and in the LWS_CALLBACK_HTTP_WRITEABLE event handler:

 goto try_to_reuse (ie, calls lws_http_transaction_completed) 

Then the server doesn't seem to 'see' the closes. It responds to 5 requests, then it sleeps for 5 seconds, then responds with 5 more and so on. 

If the LWS_CALLBACK_HTTP_WRITEABLE event handler does

return 0;

Then the server works as I would expect.

**Comment by lws-team** [comment]
That took a lot longer to understand than I expected... it's a real corner case in dealing with HTTP/1 pipelining after a POST body.  Please try current master.

**Comment by redbluetoken** [comment]
I can confirm this is fixed in master. Thanks.
